src/algebra/algebra/subalgebra.lean

```diff
@@ -148,14 +148,49 @@ instance {R : Type u} {A : Type v} [comm_ring R] [ring A] [algebra R A] (S : sub
 { neg_mem := λ _, S.neg_mem }
 
 instance : inhabited S := ⟨0⟩
-instance (R : Type u) (A : Type v) [comm_semiring R] [semiring A]
-  [algebra R A] (S : subalgebra R A) : semiring S := subsemiring.to_semiring S
-instance (R : Type u) (A : Type v) [comm_semiring R] [comm_semiring A]
-  [algebra R A] (S : subalgebra R A) : comm_semiring S := subsemiring.to_comm_semiring S
-instance (R : Type u) (A : Type v) [comm_ring R] [ring A]
-  [algebra R A] (S : subalgebra R A) : ring S := @@subtype.ring _ S.is_subring
-instance (R : Type u) (A : Type v) [comm_ring R] [comm_ring A]
-  [algebra R A] (S : subalgebra R A) : comm_ring S := @@subtype.comm_ring _ S.is_subring
+
+section
+
+/-- `subalgebra`s inherit structure from their `subsemiring` / `semiring` coercions. -/
+
+instance to_semiring {R A}
+  [comm_semiring R] [semiring A] [algebra R A] (S : subalgebra R A) :
+  semiring S := S.to_subsemiring.to_semiring
+instance to_comm_semiring {R A}
+  [comm_semiring R] [comm_semiring A] [algebra R A] (S : subalgebra R A) :
+  comm_semiring S := S.to_subsemiring.to_comm_semiring
+instance to_ring {R A}
+  [comm_ring R] [ring A] [algebra R A] (S : subalgebra R A) :
+  ring S := S.to_subring.to_ring
+instance to_comm_ring {R A}
+  [comm_ring R] [comm_ring A] [algebra R A] (S : subalgebra R A) :
+  comm_ring S := S.to_subring.to_comm_ring
+
+instance to_ordered_semiring {R A}
+  [comm_semiring R] [ordered_semiring A] [algebra R A] (S : subalgebra R A) :
+  ordered_semiring S := S.to_subsemiring.to_ordered_semiring
+instance to_ordered_comm_semiring {R A}
+  [comm_semiring R] [ordered_comm_semiring A] [algebra R A] (S : subalgebra R A) :
+  ordered_comm_semiring S := subsemiring.to_ordered_comm_semiring S
+instance to_ordered_ring {R A}
+  [comm_ring R] [ordered_ring A] [algebra R A] (S : subalgebra R A) :
+  ordered_ring S := S.to_subring.to_ordered_ring
+instance to_ordered_comm_ring {R A}
+  [comm_ring R] [ordered_comm_ring A] [algebra R A] (S : subalgebra R A) :
+  ordered_comm_ring S := S.to_subring.to_ordered_comm_ring
+
+instance to_linear_ordered_semiring {R A}
+  [comm_semiring R] [linear_ordered_semiring A] [algebra R A] (S : subalgebra R A) :
+  linear_ordered_semiring S := S.to_subsemiring.to_linear_ordered_semiring
+/-! There is no `linear_ordered_comm_semiring`. -/
+instance to_linear_ordered_ring {R A}
+  [comm_ring R] [linear_ordered_ring A] [algebra R A] (S : subalgebra R A) :
+  linear_ordered_ring S := S.to_subring.to_linear_ordered_ring
+instance to_linear_ordered_comm_ring {R A}
+  [comm_ring R] [linear_ordered_comm_ring A] [algebra R A] (S : subalgebra R A) :
+  linear_ordered_comm_ring S := S.to_subring.to_linear_ordered_comm_ring
```

Comment:
This feels like a catastrophic lack of polymorphism that is endemic to mathlib. Extending a structure doesn't give us any of its API for free without manually copying every single definition.